codalab/lib/upload_manager.py

```diff
@@ -258,3 +326,168 @@ def cleanup_existing_contents(self, bundle):
+        """
+        Get signed url for the bundle path
+        """
+        return parse_linked_bundle_url(path).bundle_path_signed_url(**kwargs)
+
+    def get_bundle_index_url(self, path, **kwargs):
+        return parse_linked_bundle_url(path).index_path_signed_url(**kwargs)
+
+
+class ClientUploadManager(object):
+    """
+    Upload Manager for CLI client. Handle file upload for CLI client.
+    """
+
+    def __init__(self, json_api_client, stdout, stderr):
+        self._client = json_api_client
+        self.stdout = stdout
+        self.stderr = stderr
+
+    def upload_to_bundle_store(
+        self,
+        bundle: Dict,
+        packed_source: Dict,
+        use_azure_blob_beta: bool,
+        destination_bundle_store=None,
+    ):
+        """
+        Bypass server upload. Upload from client directly to different blob storage (Azure, GCS, Disk storage).
+        """
+
+        # By pass server upload:
+        # 1. The server support use Azure as default storage
+        # 2. If the user specify `--store` and blob storage is on Azure
+        upload_to_disk = False
+        bundle_store_uuid = None
+        if destination_bundle_store is not None and destination_bundle_store != '':
+            storage_info = self._client.fetch_one(
+                'bundle_stores',
+                params={
+                    'name': destination_bundle_store,
+                    'include': ['uuid', 'storage_type', 'url'],
+                },
+            )
+            bundle_store_uuid = storage_info['uuid']
+            if storage_info['storage_type'] in (StorageType.DISK_STORAGE.value,):
+                upload_to_disk = True  # The user specify --store to upload to disk storage
+
+        source_ext = zip_util.get_archive_ext(packed_source['filename'])
+        if packed_source['should_unpack'] and zip_util.path_is_archive(packed_source['filename']):
+            unpack_before_upload = True
+            is_dir = source_ext in zip_util.ARCHIVE_EXTS_DIR
+        else:
+            unpack_before_upload = False
+            is_dir = False
+
+        # try to bypass server upload
+        params = {'need_sas': not upload_to_disk, 'is_dir': is_dir}
+        data = self._client.add_bundle_location(bundle['id'], bundle_store_uuid, params)[0].get(
+            'attributes'
+        )
+
+        if data.get('bundle_conn_str', None) is not None:
+            # Mimic the rest server behavior
+            # decided the bundle type (file/directory) and decide whether need to unpack
+            bundle_conn_str = data.get('bundle_conn_str')
+            index_conn_str = data.get('index_conn_str')
+            bundle_url = data.get('bundle_url')
+            bundle_read_str = data.get('bundle_read_url', bundle_url)
+            try:
+                progress = FileTransferProgress('Sent ', f=self.stderr)
+                with closing(packed_source['fileobj']), progress:
+                    self.upload_Azure_blob_storage(
```

Comment:
This function is small enough -- it's just one line -- that can we just inline the implementation of `upload_Azure_blob_storage` directly here instead?

Reply:
This fuction is small for Azure, but the upload_GCS_blob_storage function is long (since it can not use Apache Beam to upload). So I think it's better to have another function to handle the actual upload process.